test/firtool/instchoice.fir

```diff
@@ -26,6 +28,8 @@ circuit Foo:
     ; CHECK-SAME:   @FPGA -> @FPGATarget,
     ; CHECK-SAME:   @ASIC -> @ASICTarget
     ; CHECK-SAME: } (in clock: !firrtl.clock)
+    ; ASIC: hw.instance "inst" @ASICTarget(clock: %clock: !seq.clock) -> ()
+    ; DEFAULT: hw.instance "inst" @DefaultTarget(clock: %clock: !seq.clock) -> ()
```

Comment:
🤔 Making the default behavior be "replace instance choices with the default" also seems slightly odd. It seems like the default should be "do nothing". I realize that instance choice is unusable right now unless specialization of every option is run, though.

Reply:
Maybe it depends on how we define it, the default target, for an instance choice can be defined as the target when no option is specified. Also, if nothing is done here, when should the instance choice be handled in the pipeline ?

Reply:
If no default is specified, then the instance choice gets lowered to how the FIRRTL ABI defines this lowering. (This is not specified right now. However, I have one such ABI PR'd here: https://github.com/chipsalliance/firrtl-spec/pull/155.)

The underlying problem is that instance choice was only handled for this narrow, internal use case of CIRCT choosing a value for all instance choices as opposed to the more general user request for Verilog elaboration-time selection. 🥲 

The motivation here is more about aligning this with how layers are enabled and disabled, e.g., `--enable-layer=Verification` specializes away the `Verification` layer, but it leaves any other layers in place. It would seem better to have instance choice specialization work the same way, e.g., `--specialize-instance-choice=Performance=ASIC` would not affect any other instance choice in the circuit.

We do want to move in this direction. However, we don't have to move there right now with this PR.

WDYT?

Reply:
I think this makes more sense, but i'm assuming you meant to write: 

`--specialize-instance-choice=Platform=ASIC`

To specify which target is being specialized? 

Also, not to derail this (and i see there was some discussion of this in the spec PR) but it is useful to be able to specialize different instances of the same `instance_choice` to write miter-style circuits for testing the choices themselves (esp. when they are used in part of a larger circuit under test). I guess here it might just be fine to do the specialization at VCS-compile time though. For this reason i also like that default be explicitly named (as also discussed in the spec PR).

Reply:
So in the short run, we'd need a flag to achieve the effect of specializing all modules to their default and just provide that when we're passing in circuits that contain instance choices for now?

Reply:
> I think this makes more sense, but i'm assuming you meant to write: 
> 
> `--specialize-instance-choice=Platform=ASIC`
> 
> To specify which target is being specialized? 

Yes!

> it is useful to be able to specialize different instances of the same instance_choice to write miter-style circuits for testing the choices themselves (esp. when they are used in part of a larger circuit under test)

Per-instance specialization currently works. It does create a blow-up in the number of options. This is an intentional choice that complicates the design. E.g., consider instance choice used in a multicore design to replace a core with a model. The designer needs to decide what swapping flexibility they want to have. If they want to have complete flexibility to swap every core arbitrarily in an n-core design, that would have to be n-different options. They likely don't want that and will choose to have one instance choice which swaps all of the cores except one.

I'd prefer to not support arbitrary per-instance after-the-fact specialization as that amounts to mandated module duplication which then has interactions with must-dedup and designer expectations around deduplication. I'm more comfortable with this if CIRCT-mandated deduplication (must-dedup) is removed.

> So in the short run, we'd need a flag to achieve the effect of specializing all modules to their default and just provide that when we're passing in circuits that contain instance choices for now?

I'm good with either:

1. All instance choice specializations are specified up front. 
2. There is a catch-all `-specialize-to-default-if-no-override` which will, after applying selected choices, will then convert everything to the default choice.

Reply:
After more thought, i think strictness of (1) is probably desirable to avoid accidental footguns? We can always add the flag later. I'll defer to you and @prithayan on this one.

Reply:
I've decided to go with the option 1, unless all the instance choice specializations are specified, the pass will error out. 
Once the lowering of instance choice through the pipeline to verilog is supported, we can get rid of the error.

Reply:
So if you went with (1) how do you specialize to `Default`?